src/lists.py

```diff
@@ -1,90 +1,111 @@
-"""Singly-linked lists."""
-
-from __future__ import annotations
-from typing import Generic, TypeVar, Optional
-
-T = TypeVar('T')  # Generic type variable
-
-
-class Link(Generic[T]):
-    """A link in a singly linked list."""
-
-    head: T
-    tail: LList[T]
-
-    def __init__(self, head: T, tail: LList[T]):
-        """Prepend a new head to tail."""
-        self.head = head
-        self.tail = tail
-
-    def __repr__(self) -> str:
-        """Representation string."""
-        return f'Link({self.head}, {self.tail})'
-
-
-LList = Optional[Link[T]]  # A list is just a reference to the head or None
-
-
-def length(x: LList[T]) -> int:
-    """
-    Get the length of x.
-
-    >>> length(None)
-    0
-    >>> length(Link(1, None))
-    1
-    >>> length(Link(1, Link(2, None)))
-    2
-    """
-    ...
-
-
-def drop(x: LList[T], k: int) -> LList[T]:
-    """
-    Drop the first k elements in the list.
-
-    If length(x) < k, return the empty list.
-
-    >>> drop(None, 1) is None
-    True
-    >>> drop(Link(1, None), 1) is None
-    True
-    >>> drop(Link(1, Link(2, None)), 1)
-    Link(2, None)
-    """
-    ...
-
-
-def take(x: LList[T], k: int) -> LList[T]:
-    """
-    Return a list with the first k elements in x.
-
-    If length(x) < k, return the full list. You decide whether you
-    want to return a copy of x or the original list.
-
-    >>> take(None, 1) is None
-    True
-    >>> take(Link(1, None), 1)
-    Link(1, None)
-    >>> take(Link(1, Link(2, Link(3, None))), 2)
-    Link(1, Link(2, None))
-    """
-    ...
-
-
-def reverse(x: LList[T]) -> LList[T]:
-    """
-    Reverse a list.
-
-    You decide whether you are allowed to modify the existing list
-    or if you want to return a new list and leave the original one
-    intact.
-
-    >>> reverse(None) is None
-    True
-    >>> reverse(Link(1, None))
-    Link(1, None)
-    >>> reverse(Link(1, Link(2, Link(3, None))))
-    Link(3, Link(2, Link(1, None)))
-    """
-    ...
+"""Singly-linked lists."""
+
+from __future__ import annotations
+from typing import Generic, TypeVar, Optional
+
+T = TypeVar('T')  # Generic type variable
+
+
+class Link(Generic[T]):
+    """A link in a singly linked list."""
+
+    head: T
+    tail: LList[T]
+
+    def __init__(self, head: T, tail: LList[T]):
+        """Prepend a new head to tail."""
+        self.head = head
+        self.tail = tail
+
+    def __repr__(self) -> str:
+        """Representation string."""
+        return f'Link({self.head}, {self.tail})'
+
+
+LList = Optional[Link[T]]  # A list is just a reference to the head or None
+
+
+def length(x: LList[T], acc = 1) -> int:
+    """
+    Get the length of x.
+
+    >>> length(None)
+    0
+    >>> length(Link(1, None))
+    1
+    >>> length(Link(1, Link(2, None)))
+    2
+    """
+    if x is None:
+        return 0
+
+    while x.tail != None:
+        return length(x.tail, acc + 1)

```

Comment:
There are two things that are strange here.

First, the `while` loop is really an `if` statement, since if you enter it you immediately return. You never iterate more than once.

Second, you are using two base cases for the recursion, where one would suffice. You check if `x is None` (where you return 0) and then `x.tail != None` (where you return `acc` which is 1). If you just return `acc` when `x is None` and gave it the default value 0, you would get the same effect. Then you could write

```python
def length(x, acc = 0):
   if x is None: return acc
   return length(x.tail, acc + 1)
```

If you want an iterative version, it could look like this:

```python
def length(x):
   acc = 0
   while x:
      acc, x = acc+1, x.tail
   return acc
```